// angular-ssr/build/output/html.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var chalk = require("chalk");
var path_1 = require("path");
var static_1 = require("../static");
var exception_1 = require("../exception");
var filesystem_1 = require("../filesystem");
var log_1 = require("./log");
var route_1 = require("../route");
var transform_1 = require("./transform");
var HtmlOutput = (function () {
    function HtmlOutput(options) {
        this.options = options;
    }
    HtmlOutput.prototype.initialize = function () {
        if (this.options.output == null) {
            throw new exception_1.OutputException('HTML output writer needs a path to write to');
        }
        if (this.options.output.exists() === false) {
            try {
                this.options.output.mkdir();
            }
            catch (exception) {
                throw new exception_1.OutputException("Cannot create output folder: " + this.options.output.toString(), exception);
            }
            log_1.log.info("Created output path: " + this.options.output.toString());
        }
        return Promise.resolve();
    };
    HtmlOutput.prototype.write = function (snapshot) {
        return __awaiter(this, void 0, void 0, function () {
            var file;
            return __generator(this, function (_a) {
                file = filesystem_1.fileFromString(path_1.join(this.routedPathFromSnapshot(snapshot).toString(), this.options.filename || static_1.Files.index));
                transform_1.transformInplace(file.parent(), snapshot, this.options);
                log_1.log.info("Rendered route " + route_1.pathFromUri(snapshot.uri) + " to " + file);
                file.create(snapshot.renderedDocument);
                return [2 /*return*/, Promise.resolve(void 0)];
            });
        });
    };
    HtmlOutput.prototype.exception = function (exception) {
        log_1.log.error("Fatal exception encountered: " + chalk.red(exception.toString()));
        return Promise.resolve(void 0);
    };
    HtmlOutput.prototype.routedPathFromSnapshot = function (snapshot) {
        var routedPath = filesystem_1.pathFromString(path_1.join(this.options.output.toString(), route_1.pathFromUri(snapshot.uri)));
        routedPath.mkdir();
        return routedPath;
    };
    return HtmlOutput;
}());
exports.HtmlOutput = HtmlOutput;
