// angular-ssr/build/application/compiler/ngc/create.js
Object.defineProperty(exports, "__esModule", { value: true });
var compiler_1 = require("@angular/compiler");
var compiler_cli_1 = require("@angular/compiler-cli");
var path_mapped_compiler_host_1 = require("@angular/compiler-cli/src/path_mapped_compiler_host");
var tsc_wrapped_1 = require("@angular/tsc-wrapped");
var resource_resolver_1 = require("./resource-resolver");
exports.createNgCompiler = function (compilerHost, program, options, roots) {
    var host = createCompilerHost(compilerHost, roots, program, options);
    var cli = new tsc_wrapped_1.NgcCliOptions({
        i18nFormat: null,
        i18nFile: null,
        locale: null,
        basePath: options.basePath
    });
    var _a = compiler_1.createAotCompiler(host, {
        translations: null,
        i18nFormat: cli.i18nFormat,
        locale: cli.locale
    }), compiler = _a.compiler, reflector = _a.reflector;
    return { compiler: compiler, reflector: reflector, host: host };
};
var createCompilerHost = function (compilerHost, roots, program, options) {
    return roots.length > 0
        ? new path_mapped_compiler_host_1.PathMappedCompilerHost(program, options, new resource_resolver_1.ResourceResolver(compilerHost))
        : new compiler_cli_1.CompilerHost(program, options, new resource_resolver_1.ResourceResolver(compilerHost));
};
