// angular-ssr/build/application/compiler/ngc/diagnostics.js
Object.defineProperty(exports, "__esModule", { value: true });
var os_1 = require("os");
var process_1 = require("process");
var typescript_1 = require("typescript");
var exception_1 = require("../../../exception");
var transformation_1 = require("../../../transformation");
exports.diagnosticsToException = function (diagnostics) {
    var host = {
        getCurrentDirectory: function () { return process_1.cwd(); },
        getCanonicalFileName: function (filename) { return filename; },
        getNewLine: function () { return os_1.EOL; },
    };
    return "Your application failed to compile (to resolve, run tsc from " + process_1.cwd() + " and resolve these errors: " + typescript_1.formatDiagnostics(diagnostics, host) + ")";
};
exports.assertProgram = function (program) {
    exports.assertDiagnostics(program.getOptionsDiagnostics());
    exports.assertDiagnostics(program.getSemanticDiagnostics());
    exports.assertDiagnostics(program.getSyntacticDiagnostics());
    exports.assertDiagnostics(transformation_1.flatten(program.getSourceFiles().map(function (file) { return typescript_1.getPreEmitDiagnostics(program, file); })));
};
exports.assertDiagnostics = function (diagnostics) {
    diagnostics = (diagnostics || []).filter(function (d) { return d.category === typescript_1.DiagnosticCategory.Error; });
    if (diagnostics.length > 0) {
        throw new exception_1.CompilerException(exports.diagnosticsToException(diagnostics));
    }
};
