// scoped-logger/build/logger/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var level_1 = require("../level");
var Logger = (function () {
    function Logger(parent, description, additionalStreams) {
        if (additionalStreams === void 0) { additionalStreams = []; }
        this.parent = parent;
        this.description = description;
        this.additionalStreams = additionalStreams;
    }
    Logger.prototype.newScope = function (description, additionalStreams) {
        return new Logger(this, description, additionalStreams);
    };
    Logger.prototype.info = function (message) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.traverseUpward(level_1.Level.Info, this, message, args);
    };
    Logger.prototype.debug = function (message) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.traverseUpward(level_1.Level.Debug, this, message, args);
    };
    Logger.prototype.warn = function (message) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.traverseUpward(level_1.Level.Warning, this, message, args);
    };
    Logger.prototype.error = function (message) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        this.traverseUpward(level_1.Level.Error, this, message, args);
    };
    Logger.prototype.dispose = function () {
        for (var _i = 0, _a = this.additionalStreams || []; _i < _a.length; _i++) {
            var stream = _a[_i];
            stream.dispose();
        }
    };
    Logger.prototype.traverseUpward = function (level, scope, message, args) {
        for (var _i = 0, _a = this.additionalStreams || []; _i < _a.length; _i++) {
            var stream = _a[_i];
            stream.buffer(level, scope, message, args);
        }
        if (this.parent) {
            this.parent.traverseUpward(level, scope, message, args);
        }
    };
    return Logger;
}());
exports.Logger = Logger;
