// angular-ssr/build/runtime/browser-emulation/navigator.js
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var injector_map_1 = require("../../platform/zone/injector-map");
var static_1 = require("../../static");
var navigatorLanguage = null;
var navigator = {
    get appCodeName() {
        return 'Mozilla';
    },
    get appName() {
        return 'Netscape';
    },
    get appVersion() {
        return '5.0';
    },
    get cookieEnabled() {
        return false;
    },
    get languages() {
        return [navigator.language];
    },
    get language() {
        if (navigatorLanguage != null) {
            return navigatorLanguage;
        }
        var locale = injector_map_1.injectableFromZone(Zone.current, core_1.LOCALE_ID);
        if (locale == null) {
            locale = static_1.ApplicationFallbackOptions.locale;
        }
        return locale;
    },
    set language(locale) {
        navigatorLanguage = locale;
    },
    get product() {
        return 'Gecko';
    },
    get userAgent() {
        return 'Chrome';
    },
    get vendor() {
        return 'Google Inc.';
    }
};
exports.bindNavigator = function (target) { return [true, { navigator: navigator }]; };
