// angular-ssr/build/runtime/browser-emulation/storage.js
Object.defineProperty(exports, "__esModule", { value: true });
var MemoryStorage = (function () {
    function MemoryStorage() {
        this.map = new Map();
    }
    Object.defineProperty(MemoryStorage.prototype, "length", {
        get: function () {
            return this.map.size;
        },
        enumerable: true,
        configurable: true
    });
    MemoryStorage.prototype.key = function (n) {
        return Array.from(this.map.keys())[n];
    };
    MemoryStorage.prototype.setItem = function (key, value) {
        this.map.set(key, value);
    };
    MemoryStorage.prototype.getItem = function (key) {
        return this.map.get(key);
    };
    MemoryStorage.prototype.removeItem = function (key) {
        this.map.delete(key);
    };
    MemoryStorage.prototype.clear = function () {
        this.map.clear();
    };
    return MemoryStorage;
}());
exports.MemoryStorage = MemoryStorage;
var createStorage = function () {
    return new Proxy(new MemoryStorage(), {
        get: function (target, name) {
            if (name in target) {
                return target[name];
            }
            return target.getItem(name);
        },
        set: function (target, name, value) {
            if (name in target) {
                target[name] = value;
            }
            else {
                target.setItem(name, value);
            }
            return true;
        }
    });
};
exports.bindStorage = function (target) { return [true, { localStorage: createStorage(), sessionStorage: createStorage() }]; };
