// angular-ssr/build/runtime/browser-emulation/style.js
Object.defineProperty(exports, "__esModule", { value: true });
var CSSStyleDeclaration = require('domino-modernized').impl.CSSStyleDeclaration;
var getMatchedCSSRules = function () {
    return {
        matches: false,
        media: undefined,
        length: 0,
        item: function (index) { return undefined; },
        addListener: function (listener) { },
        removeListener: function (listener) { },
    };
};
var getComputedStyle = function (element) {
    var style = new CSSStyleDeclaration(element);
    var baseStyles = {
        display: 'block',
        height: 'auto',
        width: 'auto',
        'font-style': 'normal',
        'font-weight': 'normal',
        'line-height': 'normal',
    };
    for (var key in baseStyles) {
        if (style[key] == null) {
            style[key] = baseStyles[key];
        }
    }
    return style;
};
var matchMedia = function (queryString) {
    return {
        matches: false,
        media: null,
        length: 0,
        item: function (index) { return undefined; },
        addListener: function (listener) { },
        removeListener: function (listener) { },
    };
};
exports.bindStyle = function (target) { return [true, { getComputedStyle: getComputedStyle, getMatchedCSSRules: getMatchedCSSRules, matchMedia: matchMedia }]; };
