// angular-ssr/build/cache/memory-variant-cache.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var lru_map_1 = require("lru_map");
var trie_1 = require("./trie");
var MemoryVariantCache = (function () {
    function MemoryVariantCache(application, cacheSize, ttl) {
        if (cacheSize === void 0) { cacheSize = 1 << 16; }
        if (ttl === void 0) { ttl = 1 << 18; }
        this.application = application;
        this.cacheSize = cacheSize;
        this.ttl = ttl;
        this.cache = new lru_map_1.LRUMap(this.cacheSize);
    }
    MemoryVariantCache.prototype.get = function (uri, variants) {
        return __awaiter(this, void 0, void 0, function () {
            var snapshots, cached, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        snapshots = this.cache.get(uri);
                        if (snapshots == null) {
                            snapshots = new trie_1.Trie();
                            this.cache.set(uri, snapshots);
                        }
                        cached = snapshots.query(variants);
                        if (!(cached == null || (Date.now() - cached.time) >= this.ttl)) return [3 /*break*/, 2];
                        _a = {};
                        return [4 /*yield*/, this.application.renderUri(uri, variants)];
                    case 1:
                        cached = (_a.snapshot = _b.sent(),
                            _a.time = Date.now(),
                            _a);
                        snapshots.insert(variants, cached);
                        _b.label = 2;
                    case 2: return [2 /*return*/, cached.snapshot];
                }
            });
        });
    };
    MemoryVariantCache.prototype.has = function (uri, variants) {
        var snapshots = this.cache.get(uri);
        if (snapshots != null) {
            return snapshots.query(variants) != null;
        }
        return false;
    };
    return MemoryVariantCache;
}());
exports.MemoryVariantCache = MemoryVariantCache;
