// angular-ssr/build/filesystem/factories.js
Object.defineProperty(exports, "__esModule", { value: true });
var os_1 = require("os");
var path_1 = require("path");
var static_1 = require("../static");
var impl_1 = require("./impl");
exports.pathFromString = function (sourcePath) {
    if (typeof sourcePath === 'string') {
        return new impl_1.PathImpl(sourcePath);
    }
    return sourcePath;
};
exports.fileFromString = function (filePath) {
    if (typeof filePath === 'string') {
        return new impl_1.FileImpl(exports.pathFromString(path_1.dirname(filePath)), filePath);
    }
    return filePath;
};
exports.makeAbsolute = function (basePath, subpath) {
    return path_1.resolve(basePath.toString(), subpath);
};
exports.absolutePath = function (basePath, subpath) {
    return exports.pathFromString(exports.makeAbsolute(basePath.toString(), subpath));
};
exports.absoluteFile = function (basePath, subpath) {
    return exports.fileFromString(exports.makeAbsolute(basePath.toString(), subpath));
};
exports.pathFromRandomId = function () {
    var path = exports.pathFromString(path_1.join(os_1.tmpdir(), static_1.randomId()));
    path.mkdir();
    return path;
};
