// angular-ssr/build/platform/zone/environment.js
Object.defineProperty(exports, "__esModule", { value: true });
require("./assertions");
var collectors_1 = require("../collectors");
var document_1 = require("../document");
var runtime_loader_1 = require("../module/runtime-loader");
var browser_emulation_1 = require("../../runtime/browser-emulation");
var injector_map_1 = require("./injector-map");
exports.baseConsole = console;
Object.defineProperties(global, {
    console: {
        get: function () {
            return injector_map_1.injectableFromZone(Zone.current, collectors_1.ConsoleCollector) || exports.baseConsole;
        }
    },
    document: {
        get: function () { return window.document; }
    },
    history: {
        get: function () { return window.history; }
    },
    location: {
        get: function () { return window.location; }
    },
    navigator: {
        get: function () { return window.navigator; }
    },
    window: {
        get: function () {
            var documentContainer = injector_map_1.injectableFromZone(Zone.current, document_1.DocumentContainer);
            return documentContainer == null
                ? browser_emulation_1.bootWindow
                : documentContainer.window;
        }
    },
});
if (global['System'] == null) {
    Object.defineProperties(global, {
        System: {
            get: function () {
                var loader = injector_map_1.injectableFromZone(Zone.current, runtime_loader_1.RuntimeModuleLoader);
                if (loader) {
                    return { import: function (moduleId) { return loader.load(moduleId); } };
                }
                return undefined;
            }
        }
    });
}
