// scoped-logger/build/stream/file/roller-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var process_1 = require("process");
var path_1 = require("path");
var fs_1 = require("fs");
var formatter_1 = require("../formatter");
var mkdir_1 = require("./mkdir");
var rollingOptions = {
    path: process_1.cwd(),
    filenames: {
        current: 'log',
        previous: 'log.{number}',
    },
    retention: 5,
    thresholdSize: Math.pow(1024, 3),
};
var RollerStream = (function () {
    function RollerStream(options) {
        if (options === void 0) { options = rollingOptions; }
        this.options = options;
        this.options = Object.assign({}, rollingOptions, options);
        this.open();
    }
    RollerStream.prototype.buffer = function (level, scope, message, args) {
        var _this = this;
        if (this.opened() === false) {
            if (this.open() === false) {
                return;
            }
        }
        try {
            fs_1.writeSync(this.fileDescriptor, formatter_1.Formatter.simpleText(level, scope, message, args));
            if (this.rollTimer) {
                clearTimeout(this.rollTimer);
            }
            this.rollTimer = setTimeout(function () { return _this.roll(); }, 150);
        }
        catch (exception) {
            console.error("Failure to write to log file: " + exception.stack);
        }
    };
    RollerStream.prototype.dispose = function () {
        if (this.opened()) {
            fs_1.closeSync(this.fileDescriptor);
        }
        this.fileDescriptor = null;
    };
    RollerStream.prototype.opened = function () {
        return this.fileDescriptor != null
            && this.fileDescriptor >= 0;
    };
    Object.defineProperty(RollerStream.prototype, "currentPath", {
        get: function () {
            return path_1.resolve(path_1.join(this.options.path, this.options.filenames.current));
        },
        enumerable: true,
        configurable: true
    });
    RollerStream.prototype.open = function () {
        this.dispose();
        try {
            mkdir_1.recursiveCreate(this.currentPath);
            this.fileDescriptor = fs_1.openSync(this.currentPath, 'a+');
            if (this.fileDescriptor < 0) {
                throw new Error("Open returned a sub-zero file descriptor");
            }
            return true;
        }
        catch (e) {
            console.error("Failed to open log file: " + e.stack);
            return false;
        }
    };
    RollerStream.prototype.roll = function () {
        var _this = this;
        var stat = fs_1.fstatSync(this.fileDescriptor);
        if (stat.size < this.options.thresholdSize) {
            return;
        }
        var previous = this.collectPrevious();
        if (previous.length + 1 >= this.options.retention) {
            var removals = previous.filter(function (v, index) { return index + 1 >= _this.options.retention; });
            for (var _i = 0, removals_1 = removals; _i < removals_1.length; _i++) {
                var file = removals_1[_i];
                fs_1.unlinkSync(file);
                previous.pop();
            }
        }
        this.archiveCurrent(previous);
    };
    RollerStream.prototype.resolvePrevious = function (index) {
        var filename = this.options.filenames.previous.replace('{number}', index.toString());
        return path_1.resolve(path_1.join(this.options.path, filename));
    };
    RollerStream.prototype.archiveCurrent = function (previous) {
        this.shiftPrevious(previous);
        fs_1.renameSync(this.currentPath, this.resolvePrevious(0));
        this.open();
    };
    RollerStream.prototype.shiftPrevious = function (files) {
        for (var iterator = files.length - 1; iterator >= 0; --iterator) {
            fs_1.renameSync(files[iterator], this.resolvePrevious(iterator + 1));
        }
    };
    RollerStream.prototype.collectPrevious = function () {
        var iterator = 0;
        var previous = new Array();
        while (true) {
            var filename = this.resolvePrevious(iterator);
            if (fs_1.existsSync(filename)) {
                previous.push(filename);
                iterator++;
            }
            else {
                break;
            }
        }
        return previous;
    };
    return RollerStream;
}());
exports.RollerStream = RollerStream;
